File: session 4/ATM/src/ca/csf/dfc/classes/ATM.java
/**
 * 
 */
package ca.csf.dfc.classes;

/**
 * @author ManueLMaldonado
 *
 */
public class ATM {
	
	private CreateurTransaction m_createurtransaction;
	private TiroirArgent m_tiroirargent;//ok
	
	public ATM(CreateurTransaction p_createurtransaction, 
				TiroirArgent p_tiroirargent) {
		
		this.m_createurtransaction = p_createurtransaction;
		this.m_tiroirargent = p_tiroirargent;
	}
	
	
	public void Retirer(CompteCourant p_compte2, Double p_montant) {
		
		Transaction transaction = m_createurtransaction.Creertransaction(p_compte2,p_montant); 
		
		if(transaction.EstValide()) {
			
			if(!(transaction.ValiderTransaction())) {
				m_tiroirargent.DistribuerArgent();
			}
			else {
				transaction.Annuler();
			}
				
		}
		
	}
	
	public void Depositer(CompteCourant p_compte2, Double p_montant) {
		
		Transaction transaction = m_createurtransaction.Creertransaction(p_compte2,p_montant); 
		
		if(transaction.EstValide()) {
			
			if(!(transaction.ValiderTransaction())) {
				m_tiroirargent.DistribuerArgent();
			}
			else {
				transaction.Annuler();
			}
				
		}
		
	}
	
}

File: session 4/Demo solid/DemoSOLID/ATM.cs
﻿using System;
using System.Collections.Generic;
using System.Text;

namespace DemoSOLID
{

    public class ATM
    {
        private CreateurTransaction m_createurTransaction;
        private TiroirArgent m_tiroirArgent;

        public ATM(TiroirArgent p_tiroirArgent, CreateurTransaction p_createurTransaction)
        {
            m_createurTransaction = p_createurTransaction;
            m_tiroirArgent = p_tiroirArgent;
        }

        public void Retirer(Compte p_compte, decimal p_montant)
        {
            Transaction transaction = m_createurTransaction.CreerTransactionRetirer(p_compte, p_montant);
            if (transaction.EstValide())
            {
                try
                {
                    transaction.ExecuterTransaction();
                    m_tiroirArgent.DistribuerArgent(p_montant);
                }
                catch (Exception ex)
                {
                    transaction.Annuler();
                }

            }
        }
    }
}
